src/actions.ts

```diff
@@ -467,6 +475,75 @@ export const actions: {[k: string]: QueryHandler} = {
 			matches: await tables.users.selectAll(['userid', 'banstate'])`WHERE ip = ${res.ip}`,
 		};
 	},
+	// oauth is broken into a few parts
+	// oauth/page - public-facing part
+	// oauth/api/page - api part (does the actual action)
+	async 'oauth/authorize'(params) {
+		if (!params.redirect_uri) {
+			throw new ActionError("No redirect_uri provided");
+		}
+		const clientInfo = await getOAuthClient(params.client_id);
+
+		this.response.setHeader('Content-Type', 'text/html');
+		try {
+			let content = await fs.readFile(
+				__dirname + "/../../src/public/oauth-authorize.html",
+				'utf-8'
+			);
+			// table keys are owner, clientName, id
+			// expects client, client_name, redirect_uri
+			content = content.replace(/\{\{client\}\}/g, clientInfo.client_title);
+			content = content.replace(/\{\{client_name\}\}/g, clientInfo.owner);
+			content = content.replace(/\{\{redirect_uri\}\}/g, params.redirect_uri);
+			this.response.setHeader('Content-Length', content.length);
+			return content;
+		} catch (e) {
+			Server.crashlog(e, "oauth/authorize", params);
+			return "<body>The OAuth page could not be served at this time. Please try again later.</body>";
+		}
+	},
+
+	// make a token if they don't already have it
+	async 'oauth/api/authorize'(params) {
+		if (!this.user.loggedIn) {
+			throw new ActionError("You're not logged in.");
+		}
+		const clientInfo = await getOAuthClient(params.client_id);
+		const existing = await (
+			tables.oauthTokens.selectOne()
+		)`WHERE client = ${clientInfo.id} AND owner = ${this.user.id}`;
+		if (existing) {
+			if (Date.now() - existing.time > 2 * 7 * 24 * 60 * 1000) { // 2w
+				await tables.oauthTokens.delete(existing.id);
+				return {success: false};
```

Comment:
If an OAuth token is expired, shouldn't we generate another one rather than failing?

Reply:
If an oauth token is expired it's up to the client to get a new one - otherwise there's no point in expiry (which is there to be sure people don't leave something authorized forever).